homeassistant/components/hyperion/config_flow.py

```diff
@@ -0,0 +1,380 @@
+            await self.async_set_unique_id(hyperion_id)
+            self._abort_if_unique_id_configured()
+        else:
+            return self.async_abort(reason="no_id")
+        return await self.async_step_auth()
+
+    # pylint: disable=arguments-differ
+    async def async_step_user(
+        self,
+        user_input: Optional[ConfigType] = None,
+        errors: Optional[Dict] = None,
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        if not user_input or errors:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required(CONF_HOST): str,
+                        vol.Optional(CONF_PORT, default=const.DEFAULT_PORT_JSON): int,
+                    }
+                ),
+                errors=errors or {},
+            )
+
+        self._data = user_input
+        return await self.async_step_auth()
+
+    async def _cancel_request_token_task(self) -> None:
+        """Cancel the request token task if it exists."""
+        if self._request_token_task is not None:
+            if not self._request_token_task.done():
+                self._request_token_task.cancel()
+
+            # Cancellation is only processed on the **next** cycle of the event loop, so
+            # yield here.
+            await asyncio.sleep(0)
+
+            try:
+                await self._request_token_task
+            except asyncio.CancelledError:
+                pass
+            self._request_token_task = None
+
+    async def _request_token_task_func(self, auth_id: str) -> Optional[ConfigType]:
+        """Send an async_request_token request."""
+        auth_resp = {}
+        async with await self._create_client(raw_connection=True) as hyperion_client:
+            if hyperion_client:
+                # The Hyperion-py client has a default timeout of 3 minutes on this request.
+                auth_resp = await hyperion_client.async_request_token(
+                    comment=DEFAULT_ORIGIN, id=auth_id
+                )
+            await self.hass.config_entries.flow.async_configure(
+                flow_id=self.flow_id, user_input=auth_resp
+            )
+        return auth_resp
+
+    def _get_hyperion_url(self) -> str:
+        """Return the URL of the Hyperion UI."""
+        # If this flow was kicked off by SSDP, this will be the correct frontend URL. If
+        # this is a manual flow instantiation, then it will be a best guess (as this
+        # flow does not have that information available to it). This is only used for
+        # approving new dynamically created tokens, so the complexity of asking the user
+        # manually for this information is likely not worth it (when it would only be
+        # used to open a URL, that the user already knows the address of).
+        return f"http://{self._data[CONF_HOST]}:{self._port_ui}"
+
+    async def async_step_auth(
```

Comment:
I'd suggest to make this step a helper method without a form instead. It can then return different error reasons that the calling step can turn into form errors and show.

Reply:
But this step has a form it needs (static token vs create a token) -- I don't follow how that can be removed.

Perhaps you mean pull the top half of this step (where we see if auth is actually required) into a helper method, and that decides whether or not to invoke the auth step in the first place to gather the required data in the auth form?

Reply:
Yeah, isolate the action that decides the next step from the form steps.